backends/p4tools/testgen/targets/bmv2/test_spec.cpp

```diff
@@ -0,0 +1,175 @@
+#include "backends/p4tools/testgen/targets/bmv2/test_spec.h"
+
+namespace P4Tools {
+
+namespace P4Testgen {
+
+namespace Bmv2 {
+
+/* =========================================================================================
+ *  Bmv2Register
+ * ========================================================================================= */
+
+Bmv2RegisterValue::Bmv2RegisterValue(const IR::Expression* initialValue)
+    : initialValue(initialValue) {}
+
+void Bmv2RegisterValue::addRegisterCondition(Bmv2RegisterCondition cond) {
+    registerConditions.push_back(cond);
+}
+
+const IR::Expression* Bmv2RegisterValue::getInitialValue() const { return initialValue; }
+
+cstring Bmv2RegisterValue::getObjectName() const { return "Bmv2RegisterValue"; }
+
+const IR::Expression* Bmv2RegisterValue::getCurrentValue(const IR::Expression* index) const {
+    const IR::Expression* baseExpr = initialValue;
```

Comment:
We do not have the index we want to access when we push a new register condition. Is this what you mean?

Reply:
You push the register condition on a stack, and when you try to access it you build a Mux expression. Couldn't you just directly build the Mux expression here?

Reply:
No because we use the index as the condition of the argument. The read index may be different every time, but the "writes" of the register were done to a particular register index.  

For example, you could push `( index: "hdr.h.idx", val: "hdr.h.a")` into the register and later you may want to read using index 1. With the model of the function this will return  `(hdr.h.idx == 1 ? : hdr.h.a : default)`. 

You could also have a symbolic index you read from (hdr.h.jdx). Then the return value will look like  `(hdr.h.idx == hdr.h.jdx ? : hdr.h.a : default)`. 

Essentially, what we are trying to return here is the latest symbolic write to the register given a specific index.